### ISSUE TITLE
Prettier probably should not be transforming new List constructors

### ISSUE THREAD

**Issue by brianmfear** [dangmai/prettier-plugin-apex#262]
```sh
# Prettier options (if any):
```

**Input:**

```apex
// code snippet
String[] values = new String[0];
```

**Actual output:**

```apex
// code snippet
String[] values = new List<String>(0);
```

**Expected output:**

```apex
// code snippet
String[] values = new String[0];
```

**Additional information (please fill this out)**:

- OS: Windows 10
- Version: 10.0.18363.1016

**Comment by dangmai** [comment]
Hi, thank you for the bug report. Please see my reasoning for doing this [here](https://github.com/dangmai/prettier-plugin-apex/issues/115#issuecomment-524509593). It's one of the things that I don't foresee reverting unless there's a use case that I wasn't aware of. I'm going to close this ticket but feel free to add more discussions here.

**Comment by brianmfear** [comment]
@dangmai 

So, after some research, I have three main main concerns here.

First, "nobody" (in quotes, because obviously some people *do*) writes `new List<Object>(0);` I don't want this to become the standard syntax in Apex code, and I'm pretty sure most experienced devs would agree with this. It's not a natural syntax to write or read, except perhaps to some small minority. In fact, I had to double-check the docs, I'd forgotten you can write the List constructor this way, and I've been writing Apex since it was a Developer Preview. It's also not standard in any of the documentation examples. I can imagine this only causes unnecessary confusion with newer devs, too.

Second, as a code quality issue, I don't want to see `Object[] values = new List<Object>...` It creates some additional mental gymnastics that I have to go through to read this code. If you're going to change the type, do so consistently: `List<Object> values = new List<Object>(0);` would be *better* than the current situation, even if I strongly dislike the `new List<Object>(0);` notation.

Third, this flies in the face of similar conventions where the tool *doesn't* fix optional curly brackets (#74 ) because it's not a "code transformation" tool. If you're not willing to fix missing brackets, you shouldn't be willing to change the data type to something that is the least-commonly-used form just to conform to the notion that 1 AST -> 1 output. Adding missing brackets would conform to the 1 AST -> 1 output convention, Also, *other* formatters, like JavaScript, are more than happy to add brackets, parentheses, etc (at least, as an option), but there's no option here.

In summary, this is inconsistent logic at a basic level. It's okay to be opinionated, even strongly opinionated, but it should be reasonably opinionated with common coding conventions. If anything, it should be strongly opinionated in favor of the forms of code written in the developer guides, rather than some obtuse form that most developers don't use, especially seeing as how this plugin is advertised as the salesforce.com-recommended code formatter, yet doesn't adhere to the formatting used in the documentation.

**Comment by dangmai** [comment]
Thank you for your input, obviously you have strong preferences over this issue and I can sympathize with at least some of the points. I will try to answer them as best as I can, please keep in mind that I build and maintain this project by myself and I have to keep the balance between differing preferences from multiple people (including myself), and also keep it maintainable for me in the future.

- Point 1: Can I ask what you mean by "not a natural syntax"? The `new List<>()` syntax is pretty universal based on what I can see - e.g. an example from Salesforce's own code base [1], or an example from FinancialForce's very well known library fflib-apex-common [2]. If you mean the new list syntax with a set number of elements, I don't actually see either one very often "in the wild". Can you get more devs that you know to chime into this thread that strongly prefers one over the other?

- Point 2: You have a good point here, I do want to reduce the mental load for people reading the code.

- Point 3: Your point relies on the assumption that the transformations discussed in this ticket is the same as the transformation in #74, but they're not. Adding missing brackets actually changes the AST of the code, which is why doing that violates the 1 AST -> 1 output principle. The upstream Prettier project also does not add bracket when it changes AST, see [3] for example.

All in all, if you want to this change to happen, please get other developers to chime in this thread. None of the decisions made is set in stone, but I rely on developers' input for that to happen. Also, I'd appreciate a reference Pull Request just so we understand all the edge cases if such a solution is put in place (because as I explained in the other thread, your preferred way of formatting does not cover all the use cases of list initialization).

[1] https://github.com/SalesforceFoundation/NPSP/blob/96b9822b5ac851d316ff25ddb20a28feceda0855/src/classes/STG_Panel.cls#L171
[2] https://github.com/apex-enterprise-patterns/fflib-apex-common/blob/1ea4fe98eb109992b298349ab37887ca4873241b/sfdx-source/apex-common/test/classes/mocks/fflib_SObjectMocks.cls#L42
[3] https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLAZgAgBQwE4CucAlJgDpSaaRQDOEANnAHQMQDm2ZIAEnA20wB1CHgYATbsRAAaEBAAOMVNFrJQAQzx4IAdwAKWhGpQaGujQE81cgEZ4NYANZwYAZQWPUUdsnxE5OABbWzhxcTCAGQ0fAg12OAAxUSCNGGUfZBANAhgIWRAACxgghiFC1HhaTzA4N2NK1AA3SssssFobNDo4PBh9B3ZU5HQzWjg5ACtaAA8AIQdnVzcNILhI7zgRsYmQaZm3b3YmAEUCCHhthnG5TzxxvCzbDVCGAoU8bxghVHEYQuQAA4AAy3HTjIQOBRZD5wB5NLZyACO53gA0UJmytAAtFA4GEwgU8HAUahiQN4sMkKNrrtxkFUH5CHSjqdUVtqTs5DAXj8-gCkAAmbkOVAMI4AYQgQSpIDhAFYCgRxgAVF4mGk3EBNIgASSgEVgbjAnyUAEEDW4YJYmFdxgBfe1AA

**Comment by brianmfear** [comment]
@dangmai Thanks for coming back around on this.

Re: 1

I simply mean to say that `new List<Object>()` and `new Object[0]` are preferred over the slightly obtuse `new List<Object>(0)`. I'm fine with either of the first two (I prefer the array notation, but I now understand the logic behind conversion to `List<Object>`), the last one looks weird. Would it be too much to ask to exclude the 0? I feel like that's a one-line fix somewhere. Also, the example you quote is the "more" normal `List<Object>()` that I'm talking about, as opposed to `new List<Object>(0)`. I'm guessing this might be an AST limitation, again, which is really a terrible limitation, as that means I'm going to have to kill those pesky zeros myself, or maintain a personal patch for this.

Re: 2

If you can make the types consistent, I don't care which, then I'm happy (or, happier, anyways). As it stands now, I either need to regex replace after running prettier or do some pre-commit-hook shenanigans. Most of what Prettier outputs is very pretty, but I feel that this is an ugly design choice (I presume they didn't call it Prettier because it makes *ugly* code). My personal preference is the code with the fewest characters whilst being logical, but I see why not handling edge cases is a nice situation, since you can't do `String[][][]` like you can in some languages, the consistent choice would indeed be `List<List<List<String>>>`, even though it's obnoxiously longer than it needs to be. I presume this change shouldn't technically affect the AST any more than the current behavior already does.

Re: 3

So I forked the repo, just to see if I can make some quick fixes I could test; I'm pretty sure my boss would be okay with a specialty version that's less rigid than the formal specs in exchange for the freedom to implement our specific code style, preferably without reinventing the wheel (Prettier is, for the most part, a very *nice* wheel), and that's when I realized the AST change. That's unfortunate. I presumed that the AST would show a block, because it is, in my mind, technically a block (just without the brackets), but I see that it's not parsed that way. Oh well. Consider me educated, I've learned something new today about compiler design, which is a win in my book.

----

Personally, I feel that preservation of the AST should not come at the cost of code legibility. Perhaps it would make more sense to approach this from the opposite direction; the language server should be able to generate an AST where these blocks (do, if, while, for, etc) are always emitted as blocks, even when they technically are not because of the missing block syntax. Even if what Salesforce uses isn't technically the same AST as we're using to parse, as long as semantics are maintained, I think that would be a win for clean code.

----

I agree that changes shouldn't be made in a vacuum, but the fact that the same issues have been broached before means that I'm clearly not the only one that thinks the output is slightly nonsensical even though it's otherwise 99+% correct in behavior. The current behavior won't block our adoption of prettier, but it will mean extra post-processing steps that could be easily avoided. I'll ask on social media to see if any one else can weigh in on this subject.

I'll fiddle with this a bit myself, and perhaps pose a PR if I can figure out how exactly where the fixes should go. I'll also ask for feedback on social media. Thanks for the input!

**Comment by aheber** [comment]
@dangmai, would it be possible to keep this issue Open while the discussions are happening? I think closing this when encouraging discussion makes it very difficult for future readers to discover and weigh in on, it also means you're less likely to come back and continue discussions.

I also agree that `new List<Object>(0)` is never the correct output. The `0` is unneeded and unused generally. I personally use the `new List<Object>()` syntax by habit and better matching with other initializers such at `new Map<String, String>()`, I think there are some justifications for the other syntax though.

One of the arguments that comes up is that `String[] values = new String[0];` is less characters than `List<String> values = new List<String>(0);`, while that shouldn't matter it does in Apex because we're restricted in the total maximum characters available. There are some orgs that are simply forced by that limitation to adopt and enforce that formatting to be able to deploy their code. (Some people that do it anyway even if they aren't at the limit because they've been bitten before). The difference in format across a large code bases can be thousands of characters. 

This is exactly the type of behavior that Prettier should be able to enforce, stylistic choices. Knowing that there is actually a reason to format it one way or the other should weigh in on this project's decision. I think there is probably a very good reason to expose a configuration property to change this behavior. I understand some of the complexity and weight of that decision and the associated test code and long-term maintenance costs. I don't have a great "easy" answer for it. Still better than entire large and significant projects being unable to use this tool because it causes their code to be incompatible with their org.

I also support the consistent type change. If you're going to change the right then you should also change the left type.

**Comment by dangmai** [comment]
Sure, I'm happy to keep this ticket open and add a label to it.

**Comment by dangmai** [comment]
One thing that I see in you guys' posts is the complain about the `0` in the output and I'm not really sure why there's contention about that. The output has that `0` because the input has it, I can't remove it because it changes code behavior. For example:

```
String[] s = new String[0];
s[0] = 'Hello';
```
and 
```
String[] s = new List<String>();
s[0] = 'Hello';
```
Those code blocks have _very_ different behaviors, as you can try in the Dev Console. Is there some thing I'm missing out here?

**Comment by aheber** [comment]
:) (I saw your edit, just adding my quick research)

My understanding is that these are all equivalent. The reason for the `0` is that is how you initialize the array without members in that syntax. Having the `0` on the `List<String>(0)` format is redundant.

``` apex
String[] s = new String[0];
s[0] = 'Hello';
// System.ListException: List index out of bounds: 0

String[] s = new List<String>();
s[0] = 'Hello';
// System.ListException: List index out of bounds: 0

String[] s = new List<String>(0);
s[0] = 'Hello';
// System.ListException: List index out of bounds: 0

List<String> s = new List<String>();
s[0] = 'Hello';
// System.ListException: List index out of bounds: 0

String[] s = new String[];
s[0] = 'Hello';
// COMPILE ERROR: Error: Unexpected token '['.

List<String> s = new String[0];
s[0] = 'Hello';
// System.ListException: List index out of bounds: 0
```

**Comment by dangmai** [comment]
That is definitely a quirk of the platform that I'm not aware of, thanks for bringing that up :) I'm still unsure why we should make an exception for this particular case of input list with size 0, because we can't take that number out for all the cases, i.e.:

```
String[] s = new String[1];
s[0] = 'Hello';
```
and
```
String[] s = new List<String>();
s[0] = 'Hello';
```
have different behavior.

Compared that with the current implementation - keeping the list size makes sure the behavior is consistent no matter what the initial size is, I don't see the upside in adding this modification?

**Comment by brianmfear** [comment]
@dangmai It's not just a quirk, it's literally not documented. [Check it out](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_methods_system_list.htm).

The three listed constructors are:

```apex 
List<T>()
```

Creates a new instance of the List class. A list can hold elements of any data type T.

```apex
List<T>(listToCopy)
```

Creates a new instance of the List class by copying the elements from the specified list. T is the data type of the elements in both lists and can be any data type.

```apex
List<T>(setToCopy)
```

Creates a new instance of the List class by copying the elements from the specified set. T is the data type of the elements in the set and list and can be any data type.

----

Notably, the `List<T>(initialSize)` option isn't even in the documentation. You will catch developers unaware, and this is not a good thing. Use the Principle of Least Astonishment and use the documented, standard form of the constructor. The form you're advocating as the default is literally not even in the documentation as far as I can tell.

**Comment by dangmai** [comment]
When I referred to the quirk, I was talking about the fact that `new String[0]` behaves the same as `new List<String>()`, I'm not talking about the list init size, I do know those exist and Prettier Apex does support them.

I think the point about Salesforce documentation is a good one. I'm not sure why that's not documented there, because fundamentally both code blocks get parsed down to the same AST.

In the very first version of Prettier Apex I actually formatted list inits like your proposal, and I had to change it because of the implementation difficulties for the edge cases that it cannot handle. Even now I don't really want to track down all those cases myself and find workarounds for them, but as I said above, a reference pull request would be welcomed and if the implementation is maintainable I would be happy to merge it.

**Comment by dangmai** [comment]
As per @aheber point about an option for this, I can see the appeal, but I think I'll need multiple teams to chime in and say this issue is a significant barrier to them adopting Prettier before I make it an option. It really does add quite a bit of maintenance difficulty to the project for each option that is added.

**Comment by brianmfear** [comment]
@dangmai I'll try to make a PR this weekend if I can puzzle my way through it. I have a feeling I know what I'm doing, but that may just be the Dunning-Kruger effect. I'm not an expert on ASTs by a long short.

FYI, the [poll](https://twitter.com/brianmfear/status/1299064220118519813) came back with 100% for `new List<Object>();` (but a small sample size, so... take it with a grain of salt).

Still, I think this is worth pursuing, not just for me, but for all other developers out there that want clean and beautiful code.

**Comment by dangmai** [comment]
I'm looking forward to it!

The main test file you want to be looking at is [VariableDeclaration.cls](https://github.com/dangmai/prettier-plugin-apex/blob/master/tests/variable_declaration/VariableDeclaration.cls). It does not contain all the edge cases that I can think of, however. Here are a few more off the top of my head that will make formatting a challenge:

```
// These 2 lines give the exact same AST
stringArray = new String[0];
stringArray = new List<String>(0);

Integer index = 5;
// These 2 lines give the exact same AST
stringArray = new String[index];
stringArray = new List<String>(index);

// This is an example of a line that can't be formatted purely using [] syntax
List<List<String>> listOfList = new List<List<String>>();

// Similar to the line above but with list init size
List<List<String>> listOfList = new List<List<String>>(0);
```

All of the above codes are syntactically correct Apex codes that have to be handled by Prettier Apex.

**Comment by dangmai** [comment]
A few more points now that I have had time to look deeper into this issue:

- The reason for the inconsistent behavior of the left hand side `String[]` and `List<String>` not being transformed like the right side is that `jorje` actually emits different AST node for them, unlike the right hand side `new String[]` and `new List<String>()` which emit the same AST (NewListInit). So it comes back to the fact that if we transform the left hand side, we actually change the AST.

- Let me delve a bit into why not changing the AST is so important for this project. First off, I hope it's clear that as long as we can keep the AST unchanged after formatting, it's guaranteed that the code behavior has not changed. Because of this simple property, we can do a lot of things to make sure Prettier Apex is correct, for example:

  + We can support Prettier's `debug-check` out of the box [1].
  + As part of the unit test suite, we can do a comparison of the AST before and after the format to make sure they match, see [2].
  + The integration test suite runs Prettier on multiple open source Apex code bases, see [3]. Again, we compare the AST before and after the format to make sure they match.

- These test suites are integral to the project and have saved me many many times from making mistakes when printing out formatted code without understanding all of its consequences. It's so important that it's part of the CI/CD pipeline, which means I can't release a new version without both test suites passing. 

- However, this doesn't mean that changing the AST if out of the question. _If_ we make changes to the AST as part of the parsing phase instead of the printing phase, then it won't affect the test suites. We do need to make sure that those changes are 100% correct however - the highest priority for me in this project is correctness. For example, here's one possibility: as part of the parsing phase, we can find a way to differentiate `new List<String>()` and `new String[]` and assign them to different node types; or group the left hand side `List<String>` and `String[]` as the same node type. The difficult thing is how to parse all of those correctly. Putting in another parsing engine beside jorje into this library is a lot of work for relatively small return, which is why I bring up maintainability in my earlier post.

[1] https://prettier.io/docs/en/cli.html#--debug-check
[2] https://github.com/dangmai/prettier-plugin-apex/blob/2f76d7e27489a2d40db0d8be19c5fb60aaef8d70/tests_config/run_spec.js#L85
[3] https://github.com/dangmai/prettier-plugin-apex/blob/master/tests_integration/test-repos.txt

**Comment by brianmfear** [comment]
@dangmai After learning a bit of AST, and reading your comments, I came up with a compromise that I think most developers can live with. This change does not change the AST (so, I can't `new List<Object>()` as a sub for `new List<Object>(0)`), but it does change it to the arguably better `new Object[size]` syntax when we're certain it is safe to do so. I'd like to learn more about the AST and propose additional changes, but this one change, I feel, makes the code read much more like how I, as a human being who writes code, would like to write my code, and how I'd prefer to read it. 

As a second step, I would love to fix `new List<Object>(dynamicSize)` to `new Object[dynamicSize]`, but I couldn't figure out how I'd detect that easily without running afoul of the `new List<Object>(otherList)` syntax. I'll need to do more learning, but I loved an opportunity to finally contribute something, no matter how small, to such a useful projeect.

**Comment by dangmai** [comment]
Hey Brian, thank you for the PR, I appreciate it. I'll get to review it some time this week, but from a quick skim it looks good.

**Comment by aheber** [comment]
@dangmai, as a bridge, would it make sense to run a simpler path in the printing phase and have conditional output?

The possibility here being to declared variable type and if they are assigning to an array, then print as an array. If they are assigning to a typed list (not knowing the exact AST identifiers) then print it in that style. I realize the rule isn't entirely that clean but could be close. (e.g. If it doesn't have a numbered initializer then it must be printed in List form)

This does print different output for the same AST node, but if you can infer "correct" or preferred output based on assignment then maybe that is OK? I too am trying to get my head around the rules and guidelines here. This would isolate the change to a smaller part of the code and would still have a consistent output that I believe could support robust tests, just influenced by the world around it instead of printed in isolation. This wouldn't change the generated AST and would preserve some small opinions from the writer without surprising them.

If a code base wants to enforce one over the other then assume that needs to be a PMD or other linter rule and move the responsibility away from Prettier as that is not its purpose (if I'm understanding correctly) but this change would also make sure Prettier could be used in this case and not just turned off.

Maybe this is a terrible idea, feel free to tell me all the ways it might be :)

**Comment by dangmai** [comment]
Hey @aheber, sorry it took me a while to get back to you about this. What you're proposing has essentially the same outcome as my comment ealier, although we go about it a bit differently.

The crux of the issue is that jorje gives us the same AST for both `new List<Object>()` and `new Object[]`, and we are trying to print different things based on the user input instead of relying on the AST given to us by jorje. If I understand correctly, what you're saying is that we should look at the user input, and from there print different things depending on what form they put in; and we should do it the printing phase.

The thing is that, "looking at the user input" is essentially "parsing the user input" - hence my earlier comment about moving that part to the _parsing_ phase. Basically, we'll do a 2 phase parsing - first one is directly from jorje, then another past to figure out whether the user originally put in `new List<Object>()` or `new Object[]`, and change the node type altogether (maybe one of them can be `NewListInit` and the other one can be `NewLiteralListInit`). From there we can have different printer for those 2 node types. The end result will be the same as what propose, but each part will be contained in its correct phase.

In any case, for now I prefer to keep this issue just about the list initializer with literal integer param, because the current way of formatting is not documented by Salesforce. For the bigger change about keeping user input style of initializing lists, feel free to open another issue and we can discuss about the correct approach there.

**Comment by MorganMarchese** [comment]
Hi all,

I work for a very large org with a lot of code today and came across this surprisingly thorough post, so thanks for that. I'm curious if this ever went anywhere? I also typically use the Object[0] format and am wondering if the plugin still handles it this way or if something was changed?

**Comment by dangmai** [comment]
Hi @MorganMarchese, the behavior hasn't changed since my last post. I put in a few comments on Brian's PR, and when those are addressed I can merge it. Please note that the PR only addresses a portion of the OP.

**Comment by brianmfear** [comment]
@MorganMarchese Hi, thanks for bringing this to my attention. I missed the notifications for the review from... wow, like almost half a year ago. I thought I was waiting on @dangmai, and so I sincerely apologize. I made the requested changes, it should probably be good to go now. I'm also working on my own personal parser that I expect to have out in April or so after having recovered from some surgeries recently.

The PR I made was constrained by Prettier's general rules regarding changing the AST (as in, don't do it), and the changes I wanted to make would have violated that principle. As such, I've taken it about as far as it can go safely without violating the AST rules. However, I feel it also probably covers the most common use cases, and like @dangmai, I don't think we should be trying to focus too hard on edge cases without careful consideration. I may give this another attempt in the near future, armed with some knowledge I gained from this PR.
